### ISSUE TITLE
Add project roadmap

### ISSUE THREAD

**Issue by hferentschik** [minishift/minishift#207]
The roadmap should describe the overall vision and the main short term design decision.

**Comment by thatdocslady** [comment]
@hferentschik @LalatenduMohanty @Preeticp I'd like to suggest the following breakdown to start with (based on what we discussed last week):

# Before 1.0 release
- Clean up and strengthen existing doc set
- Prepare publication workflow
- No new/expanded content will be added.

# After 1.0 release
- Fix doc bugs and improve doc set based on whatever issues are opened as needed
- Determine top major use-cases or scenarios that we want to document end-to-end tutorials for. We can/should consult with users who are implementing the project to see which scenarios rise up to the surface
- Collaborate with OpenShift to improve integration of docs with their getting started docs as needed

**Comment by hferentschik** [comment]
There might be a misunderstanding, this issue is about adding a ROADMAP.md to the Minishift _docs_ to describe the roadmap of Minishift in terms of major features, etc. This is not a task for just planning the work on the documentation per se.

**Comment by thatdocslady** [comment]
@hferentschik indeed this is a misunderstanding, since the title of this issue is "roadmap *for* docs", and the description doesn't clarify. Since I actually think we need to document our docs roadmap/plan somewhere, maybe we can keep this issue for that and create a separate issue for the project roadmap, and name it something like "Add *project* roadmap *to* docs", to differentiate.

**Comment by thatdocslady** [comment]
I will create a separate issue to track docs roadmap and take my notes from here to there :)

**Comment by hferentschik** [comment]
> I will create a separate issue to track docs roadmap and take my notes from here to there :)

Ok

**Comment by hferentschik** [comment]
To be clear on what this is about. We are talking about a ROADMAP.md which on a more abstract levels covers the major goals of Minishift moving forward. A lot of this is covered in the milestones, but the idea is to provide some high level overview probably also paired with some sort of vision beyond the current milestone issues/features.

**Comment by jorgemoralespou** [comment]
@hferentschik Since there is no Roadmap, I will add some comments to this page that will be helpful for you to create one. Feel free to discuss the topics.

- Multiple profiles/clusters that can be started/stopped
- Ability to easily select the version of the cluster to run (to test with different versions)
- Ability to bootstrap a cluster (add templates, imagestreams)
- Proxy management. Be able to set up a proxy.
- Volume management command. Create volumes of different types. By default "oc 1.5 will provide 100 PVs with Recycle policy and 3 accessmodes of 100GB of local storage)
- User management. Configure htpasswd as IdentityProvider and add commands to adduser,removeuser,set/changepassword
- Add label to buildConfig so that images can be identified as created by minishift/CDK for provenance and cleanup
- Being able to use you own certificates for the cluster
- Being able to use LetsEncrypt for creating the certificates
- Allow to provide a hostname and domainName
- Have sane (configurable) default options to apply to all created clusters
- snapshot/restore the cluster

Some options will be easier with newer versions of oc since 1.5 has made big progress, but still far in the horizon.

These are things that we either have or will have as we consider them very important for developers.
Implementations can be seen in:
https://github.com/openshift-evangelists/oc-cluster-wrapper
or
https://github.com/getwarped/powershift-cluster

**Comment by hferentschik** [comment]
> Since there is no Roadmap, I will add some comments to this page that will be helpful for you to create one. Feel free to discuss the topics.

@jorgemoralespou, awesome. Much appreciated. I had a quick look at the list and a lot sounds familiar and is either already tracked by dedicated issues or at least in discussion within the team. Others are new to me. For example I have so far not thought about the need/possibility to bring your own certificates. Nor do I know LetsEncrypt. Sounds reasonable though.

I will take your ideas onboard for sure. I know this task has been sitting here for a while. I am sorry it ha not happened yet. Partly this is based on the fact that we are still dealing with ground work to provide the platform to add all these features. Initially I might comment on some of the ideas on this issue to get some more feedback from you. Your in depth experience with actually using what Minishift provides is super important to us. You are to a certain degree ahead of us by knowing what customisations are needed on top of plain OpenShift to provide a useful environment for our users. Really here lies the big value proposition of Minishift compared to just using 'oc cluster up' directly.

**Comment by jorgemoralespou** [comment]
@hferentschik Sure. Keep in mind that I represent not only the OpenShift Evangelist team but also the OpenShift PM team, and so I have double intentions and goals.

Also, I've noted that there's changes happening in "oc cluster" that you're not aware and that will introduce some hiccups in your tool, and should be considered. I'm working on a regular basis with @csrwng to make sure I understand all the changes we're doing in oc cluster to be able to work with you so you can take them into account, as I do (and @grahamdumpleton) in the wrappers we're creating.

Things to note:
origin 1.4 has
-- proxy support
-- volume support with 100 PVs provisioned by default with Recycle policy and all accessmodes enabled. These use host persistence if configured, or otherwise will persist on the origin container. Also with this change, the registry and metrics/logging are persistent.
-- ocp binary checks in access.registry for images while origin binary checks in github

ocp 3.4 has not the first 2 features but it has the last one. We expect the next ocp version, 3.4.1 to contain the proxy and volume support, so that CDK can leverage those. (@csrwng take this into account).

The idea is that many of the features described in my first comment can be achieved right now with the current versions, as we've proved in the wrapper scripts. Any new requirement needs to be evaluated where will be implemented, if minishift or oc cluster, but the idea is to keep oc cluster responsibilities to a minimum while providing support to greater tools (wrappers/minishift) to solve those bigger problems.

There's some use-cases that minishift will have to do differently than the wrappers just because it's using virtualization. Namely the pull down and clean up of images. In the wrappers, what we need is to be able to remove all the images that are built in a cluster when removed, but all the images that are pulled for one cluster are available to any other as live in the local docker daemon. In minishift is the opposite, a cluster removal will delete every image built within that cluster, but introduces the need to cache images so that pulling these down from multiple clusters don't introduce too much overhead.

This is the main and probably only big difference I see.

**Comment by csrwng** [comment]
Just a clarification... OCP 3.4.1 should include proxy support, but not automatically created volumes. That will come in OCP 3.5.

**Comment by GrahamDumpleton** [comment]
FWIW. My testing of oc 1.5 shows that volumes are not added by default (unless it takes the background task some time to create them). You still need to provide an option to specify where on the container file system or host file system they should go before anything happens. Thus you don't need to use the builtin functionality and could ignore it and do it yourself as we currently do in our wrappers, so that you can control number of volumes and how they are configured.

My testing also seems to have uncovered an issue with ``openshift ex config patch`` command run inside of a container. If MiniShift was going to use that to patch the master config file, there is a chance could trip the same problem depending on what you are doing. The issue appears to be in upstream code, so don't know at this point when/if might be fixed or whether need to start doing yaml parsing in the wrapper to make changes to master config file.

**Comment by hferentschik** [comment]
> If MiniShift was going to use that to patch the master config file

That's the current plan

> My testing also seems to have uncovered an issue with openshift ex config patch command run inside of a container.

What is the issue? Can you provide some more context?

> The issue appears to be in upstream code, so don't know at this point when/if might be fixed 

If we can pinpoint the problem and it is a bug upstream, I think it is ok to still go with 'openshift ex config' and potentially live with the issue until it is fixed upstream. For me the 'openshift ex config' is still the most appealing solution atm.

**Comment by GrahamDumpleton** [comment]
@csrwng can possibly explain the ``openshift ex config patch`` problem as although I got the error he confirmed it as an issue and so would better understand it. All I know is that when I tried to patch build config defaults to add labels to all built images, in case where proxy config had already been added, it would fail. All I understand so far is that it is doing something wrong with ``RawExtension`` validation when patching.

**Comment by csrwng** [comment]
@hferentschik so the issue is that `openshift ex config patch` fails if you're trying to patch the BuildConfigDefaults after it has been added to the master-config.yaml. The reason this happens is because the BuildDefaults config, like any other plugin config is of type RawExtension (https://github.com/openshift/origin/blob/master/pkg/cmd/server/api/v1/types.go#L1235). This means that for the patch code, this looks like a blob of stuff. It can set it the first time, but it can't modify it afterwards. 

Here's @GrahamDumpleton's example. If you start from this config:
```
admissionConfig:
 pluginConfig:
   BuildDefaults:
     configuration:
       apiVersion: v1
       env:
       - name: https_proxy
         value: http://localhost:8080
       - name: NO_PROXY
         value: 0.0.0.0/0,127.0.0.1,localhost,172.30.1.1,172.30.0.0/8,192.1
68.65.2
       - name: no_proxy
         value: 0.0.0.0/0,127.0.0.1,localhost,172.30.1.1,172.30.0.0/8,192.1
68.65.2
       - name: HTTP_PROXY
         value: http://localhost:8080
       - name: http_proxy
         value: http://localhost:8080
       - name: HTTPS_PROXY
         value: http://localhost:8080
       gitHTTPProxy: http://localhost:8080
       gitHTTPSProxy: http://localhost:8080
       gitNoProxy: 0.0.0.0/0,127.0.0.1,localhost,172.30.1.1,172.30.0.0/8,19
2.168.65.2
       kind: BuildDefaultsConfig
       resources: {}
     location: “"
```

And try to apply a patch:
```
#!/bin/bash
PATCH=`cat <<EOF
{
   "admissionConfig": {
       "pluginConfig": {
           "BuildDefaults": {
               "configuration": {
                   "apiVersion": "v1",
                   "kind": "BuildDefaultsConfig",
                   "imageLabels": [
                       {
                           "name": "powershift-profile",
                           "value": "proxy"
                       }
                   ]
               }
           }
       }
   }
}
EOF`

cp /var/lib/origin/openshift.local.config/master/master-config.yaml /tmp/master-config.yaml

openshift ex config patch /tmp/master-config.yaml     --patch "$PATCH" > /var/lib/origin/openshift.local.config/master/master-config.yaml
```

The patch operation fails. This will be the case with any admission plugin config.

**Comment by thesteve0** [comment]
While @jorgemoralespou has a nice list there are certain pieces I would like to prioritize so we can replace the all in one VM with minishift. My list of requirements are much smaller and hopefullly could be rolled out sooner. 

We have deprecated this due to so many problems with virtualbox and vagrant
https://www.openshift.org/vm/

We have basically stopped using the all-in-one for workshops so those requirements can be deprioritized. 

To just replace our current system we just need:

1. Pick a tag from origin releases to use to build the "cluster"
2. Ability to change the amount of memory, disk space, and CPUs dedicated to the "cluster"
3. Ability to mount PVs and PVCs in the "cluster"
4. Ensure configurations, changes 
5. [Optional] Ability to specify docker images to preload into the machine
6. [Optional] It would be great if there was an easy way to package up customs "configurations" and make it easy to share them.

**Comment by budhrg** [comment]
@thesteve0  Thanks.

> While @jorgemoralespou has a nice list there are certain pieces I would like to prioritize so we can replace the all in one VM with minishift. 

Good idea. :+1:  

> We have deprecated this due to so many problems with virtualbox and vagrant
https://www.openshift.org/vm/

We faced similar issues too.

> Run on CentOS

One can use `--iso-url` to refer [centos-iso](https://github.com/minishift/minishift-centos-iso/releases/tag/v1.0.0-beta.1).

> Pick a tag from origin releases to use to build the "cluster"

Need to plan

> Ability to change the amount of memory, disk space, and CPUs dedicated to the "cluster"

Provided by `minishift start` [options](https://github.com/minishift/minishift/blob/master/docs/minishift_start.md)

> Ability to mount PVs and PVCs in the "cluster"

Doesn't have support ATM and on future roadmap however we have [Persistent Volume of `hostPath` support](https://github.com/minishift/minishift/blob/master/docs/using.md#persistent-volumes)

> Ensure configurations, changes

We do have [Persistent configuration](https://github.com/minishift/minishift/blob/master/docs/using.md#persistent-configuration).

**Comment by hferentschik** [comment]
I am going to copy some of the _openshift ex config patch_ comments to issue #276 where they belong ...

**Comment by hferentschik** [comment]
@jorgemoralespou 

> Also, I've noted that there's changes happening in "oc cluster" that you're not aware and that will introduce some hiccups in your tool, and should be considered.

Do you have some pointers on what you mean? 

> I'm working on a regular basis with @csrwng to make sure I understand all the changes we're doing in oc cluster

I think it makes sense to be in this loop as well. Do you have a regular meeting or any other form in which you share what's happening in this area?

**Comment by hferentschik** [comment]
> Multiple profiles/clusters that can be started/stopped

This one is probably one of the hardest ones. Minishift is really geared towards single instance right now. A lot of things are hard wired here and it won't be super easy to change. Also commands needs adjusting as well. Most commands will either need some sort of '--name' option similar to docker-machine or one needs a way to select the active cluster at which stage all commands apply to this active cluster until the selection is changed.

> Ability to easily select the version of the cluster to run (to test with different versions)

+1, we are working on this. It is not properly working atm, but the main pieces should be in place.

> Ability to bootstrap a cluster (add templates, imagestreams)

+1 Work in progress. See also https://github.com/minishift/minishift/issues/177

> Proxy management. Be able to set up a proxy.

+1 one of the highest priorities atm. Will be part of the next sprint - https://github.com/minishift/minishift/issues/90. We will piggyback on 'oc cluster' for this

> Volume management command. Create volumes of different types. By default "oc 1.5 will provide 100 PVs with Recycle policy and 3 accessmodes of 100GB of local storage)

+1 We will need to create an issue for that.

> User management. Configure htpasswd as IdentityProvider and add commands to adduser,removeuser,set/changepassword

+1 Issue needs to be created

> Add label to buildConfig so that images can be identified as created by minishift/CDK for provenance and cleanup

@jorgemoralespou, can you explain this one a bit more and what needs to be done here from a provisioning/Minishift side

> Being able to use you own certificates for the cluster

+1 As said, sounds reasonable. Issue needs to be created.

 > Allow to provide a hostname and domainName

You mean for routing? Technically you can already set the appropriate flag, right? The problem is more how to make this routes work then on a local dev machine. See https://github.com/minishift/minishift/issues/201

> snapshot/restore the cluster

@jorgemoralespou, what do you have in mind there?

**Comment by GrahamDumpleton** [comment]
The addition of labels to images built in the cluster is only relevant where supporting multiple profiles. Labelling the images provides a way of being able to delete the images from the local Docker service that ``oc cluster up`` was run against, when the profile is destroyed. I don't believe it is relevant to minishift since you are only going to have the one instance with no concept of profiles.

**Comment by jorgemoralespou** [comment]
@hferentschik Please align with minikube for the plugins/addons, at least on the syntax, even if the code is not the same: https://github.com/kubernetes/minikube#add-ons  https://github.com/kubernetes/minikube#configuring-kubernetes

**Comment by hferentschik** [comment]
Merged via pull request #393.
